Apply some hlint suggestions, silence some others.

peterwicksstringfield opened the pull request:
These suggested changes come from running hlint on the entire code base, using the configuration at ghcide/.hlint.yaml

```
~/haskell-language-server$ hlint --report --with-group=extra --hint=ghcide/.hlint.yaml --ignore-glob='test/testdata/**' --ignore-glob='test/wrapper/testdata/**' --ignore-glob='ghcide/test/data/**' --ignore-glob='plugins/hls-eval-plugin/test/testdata/**' .
```

If we plan to expand hlint coverage across the code base we will need to apply or silence all of these at some point.

cf. #693

peterwicksstringfield commented on hie-compat/src-ghc86/Compat/HieBin.hs:
```diff
@@ -327,6 +327,9 @@ putName (HieSymbolTable next ref) bh name = do
+-- | Panics if the name is of sort Internal or System.
 toHieName :: Name -> HieName
 toHieName name
   | isKnownKeyName name = KnownKeyName (nameUnique name)
```
nameModule vs nameModule_maybe. Presumably we trust this usage since it's here in the code and working.

Personally I would favor defining and using (unsafeFromJust $ nameModule_maybe name). Or unsafeNameModule.

peterwicksstringfield commented on plugins/hls-eval-plugin/src/Ide/Plugin/Eval/Code.hs:
```diff
@@ -80,6 +80,8 @@ asStmts (Example e _) = NE.toList e
+
 -- |Evaluate an expression (either a pure expression or an IO a)
 evalExpr :: GhcMonad m => [Char] -> m String
 evalExpr e = do
```
Line 89 unsafeCoerce.

ndmitchell replied:
We should add the exceptions to this in .hlint.yaml, so we can audit the list if we ever want

peterwicksstringfield commented on plugins/hls-eval-plugin/src/Ide/Plugin/Eval/Parse/Token.hs:
```diff
@@ -145,6 +145,8 @@ tokens = concatMap (\(l, vs) -> map (Located l) vs) . zip [0 ..] . reverse . snd
+
 -- | Parse a line of input
 aline :: State -> TParser
 aline InCode          = optionStart <|> multi <|> singleOpen <|> codeLine
```
Line 142.

I don't think this is an improvement but we can do whatever.

plugins/hls-eval-plugin/src/Ide/Plugin/Eval/Parse/Token.hs:142:55-64: Warning: Use zipFrom
Found:
  zip [0 .. ]
Perhaps:
  zipFrom 0

peterwicksstringfield commented on plugins/hls-splice-plugin/hls-splice-plugin.cabal:
```diff
@@ -18,6 +18,7 @@ library
+                  , extra
                   , foldl
                   , haskell-lsp
                   , hls-plugin-api
```
Alternatively: don't incur a dependency just to satisfy hlint's obsession with eitherM. But the rest of hls already depends on extra so why not?

ndmitchell replied:
HLint only suggests extra functions if you pass the appropriate option, so if you don't want to use more option, then don't pass that.

peterwicksstringfield commented on plugins/tactics/src/Ide/Plugin/Tactic.hs:
```diff
@@ -192,19 +193,19 @@ provide tc name _ plId uri range _ = do
-    False -> pure []
+  if xopt ext dflags
+    then tp dflags plId uri range jdg
+    else pure []
```
The use of "case ... True ... False" is consistent and obviously deliberate in the tactics plugin. Should we just ignore the use case suggestion for this module and leave the code alone?

(I don't care either way, I basically made an arbitrary decision just to get the PR started.)

wz1000 commented:
Please don't change anything in `hie-compat`, we want the source to be as close to the original GHC version as possible.

peterwicksstringfield commented on plugins/tactics/src/Ide/Plugin/Tactic/Context.hs:
```diff
@@ -33,6 +36,9 @@ mkContext locals tcg = Context
   }
```
Right now this function is one operation per line. The suggested changes would combine some of the lines, which seems worse.

plugins/tactics/src/Ide/Plugin/Tactic/Context.hs:(29,22)-(31,30): Suggestion: Use <=<
Found:
  (getFunBindId =<<) . fmap unLoc . bagToList
Perhaps:
  (getFunBindId Control.Monad.<=< (fmap unLoc . bagToList))

Ailrun replied:
Maybe
```
((getFunBindId . unLoc) <=< bagToList)
```
is better?
I'm not super inclined to this though...

peterwicksstringfield replied:
I could move these:
```
{- HLINT ignore mkContext "Redundant fmap" -}
{- HLINT ignore mkContext "Use <=<" -}
```

to the .hlint.yaml?

Ailrun replied:
I think, `Redundant fmap` rule may be useful for some case....

peterwicksstringfield replied:
I moved the fish recommendation ignore to the hlint.yaml and left the fmap recommendation in the module.

peterwicksstringfield commented on plugins/tactics/test/AutoTupleSpec.hs:
```diff
@@ -48,10 +48,11 @@ spec = describe "auto for tuple" $ do
 randomGroups as = do
   n <- choose (1, length as)
   (:) <$> pure (take n as)
       <*> randomGroups (drop n as)
```
plugins/tactics/test/AutoTupleSpec.hs:55:3-26: Warning: Redundant <$>
Found:
  (:) <$> pure (take n as)
      <*> randomGroups (drop n as)
Perhaps:
  pure ((:) (take n as))
      <*> randomGroups (drop n as)

Not an improvement.

maralorn replied:
Wouldn’t `(take n as:) <$> randomGroups (drop n as)` work?

peterwicksstringfield replied:
Changed, thanks!

peterwicksstringfield commented on test/utils/Test/Hls/Util.hs:
```diff
@@ -126,6 +126,8 @@ ignoreForGhcVersions vers reason
+
 -- | The command to execute the version of hls for the current compiler.
 --
 -- Both @stack test@ and @cabal new-test@ setup the environment so @hls@ is
```
Line 137.

Alternatively: do add NOINLINE pragma. Shouldn't really matter, getting the name of the exe should be effectively pure.

ndmitchell replied:
It might be effectively pure, but it also might not be that cheap, causing it to be evaluated multiple times. This is definitely one that should be fixed.

peterwicksstringfield commented:
The failing build is this thing:
```
       but got: "{- |\nTests in plain comments in both single line or multi line format, both forward and backward.\nTests are ignored if:\n    * do not start on the first column\n    * are in multi line comments that open and close on the same line\n-}\nmodule THaddock () where\n\n{- ORMOLU_DISABLE -}\n\n-- | Single line comment\n-- >>> \"a\"++\"b\"\n-- <BLANKLINE>\n-- ByteCodeLink.lookupCE\n-- During interactive linking, GHCi couldn't find the following symbol:\n--   interactive_Ghci1_asPrint_closure\n-- This may be due to you not asking GHCi to load extra object files,\n-- archives or DLLs needed by your current session.  Restart GHCi, specifying\n-- the missing library using the -L/path/to/object/dir and -lmissinglibname\n-- flags, or simply by naming the relevant files on the GHCi command line.\n-- Alternatively, this link failure might indicate a bug in GHCi.\n-- If you suspect the latter, please report this as a GHC bug:\n--   https://www.haskell.org/ghc/reportabug\n\n{- | Multi line comment\n\n>>> \"b\"++\"c\"\n<BLANKLINE>\nByteCodeLink.lookupCE\nDuring interactive linking, GHCi couldn't find the following symbol:\n  interactive_Ghci1_asPrint_closure\nThis may be due to you not asking GHCi to load extra object files,\narchives or DLLs needed by your current session.  Restart GHCi, specifying\nthe missing library using the -L/path/to/object/dir and -lmissinglibname\nflags, or simply by naming the relevant files on the GHCi command line.\nAlternatively, this link failure might indicate a bug in GHCi.\nIf you suspect the latter, please report this as a GHC bug:\n  https://www.haskell.org/ghc/reportabug\n-}\n\ndouble a = a + a\n-- ^ Single line backward comments\n-- >>> double 11\n-- <BLANKLINE>\n-- ByteCodeLink.lookupCE\n-- During interactive linking, GHCi couldn't find the following symbol:\n--   interactive_Ghci1_asPrint_closure\n-- This may be due to you not asking GHCi to load extra object files,\n-- archives or DLLs needed by your current session.  Restart GHCi, specifying\n-- the missing library using the -L/path/to/object/dir and -lmissinglibname\n-- flags, or simply by naming the relevant files on the GHCi command line.\n-- Alternatively, this link failure might indicate a bug in GHCi.\n-- If you suspect the latter, please report this as a GHC bug:\n--   https://www.haskell.org/ghc/reportabug\n\ntwice :: [a] -> [a]\ntwice a = a ++ a\n{- ^ Multi-line backward comments\n>>> twice \"ABC\"\n<BLANKLINE>\nByteCodeLink.lookupCE\nDuring interactive linking, GHCi couldn't find the following symbol:\n  interactive_Ghci1_asPrint_closure\nThis may be due to you not asking GHCi to load extra object files,\narchives or DLLs needed by your current session.  Restart GHCi, specifying\nthe missing library using the -L/path/to/object/dir and -lmissinglibname\nflags, or simply by naming the relevant files on the GHCi command line.\nAlternatively, this link failure might indicate a bug in GHCi.\nIf you suspect the latter, please report this as a GHC bug:\n  https://www.haskell.org/ghc/reportabug\n-}\n\n{- | >>> 2+five\n7\n\n   ^-- This works, as it starts at the first column after the header.\n\n     >>> IGNORED as it does not start on the first column\n-}\nfive = 5\n"
```

Ailrun commented:
@peterwicksstringfield It's eval plugin's non-deterministic bug, and it may be gone after a few trials.

Ailrun commented:
@peterwicksstringfield Could you rebase this? Now the eval subtlety has (presumably) fixed, and it will be better to merge (or close) this PR before sending a style formatter PR.

peterwicksstringfield commented:
Ailrun: happy too. Wasn't sure what the state of the this PR was and wasn't sure how to ask.

Will push in an hour or so.

Ailrun commented on plugins/hls-eval-plugin/src/Ide/Plugin/Eval/Types.hs:
```diff
@@ -55,6 +55,8 @@ type Loc = Located Line
 type Line = Int
 
+{- HLINT ignore locate "Use zipWithFrom" -}
+
```
If we do not use the rule for code like this, then will we ever use this rule?
If we won't, I think it's better to put this in the hlint setting.

peterwicksstringfield replied:
You're right. I'll just apply this hint too.

Ailrun commented on plugins/hls-tactics-plugin/src/Ide/Plugin/Tactic/Judgements.hs:
```diff
@@ -89,6 +89,8 @@ withNewGoal :: a -> Judgement' a -> Judgement' a
 
 
+{- HLINT ignore introducing "Use zipFrom" -}
+
```
Same for here

Ailrun commented on plugins/hls-tactics-plugin/src/Ide/Plugin/Tactic/Judgements.hs:
```diff
@@ -218,6 +218,8 @@ filterSameTypeFromOtherPositions dcon pos jdg =
 
 
+{- HLINT ignore getAncestry "Replace case with maybe" -}
+
```
Can I ask why this is disabled?

peterwicksstringfield replied:
I don't remember what I was thinking. The version with maybe looks fine to me now. Fixed.

Ailrun commented on stack.yaml:
```diff
@@ -1,4 +1,4 @@
-resolver: lts-14.27 # Last 8.6.5
+resolver: nightly-2020-12-09
```
I wonder why did you update this?

peterwicksstringfield replied:
Merge error. I very carefully checked to make sure that that wasn't included before I pushed. Apparently I checked wrong.

Ailrun reviewed (approved):
LGTM

peterwicksstringfield commented:
Thanks for the help. :)

Ailrun commented:
Could you rebase this PR?

peterwicksstringfield commented:
Meanwhile, on the other side of the code base: 
https://github.com/haskell/haskell-language-server/pull/1391/commits/529e4dd229c7eda10c3eaa6b1613ac69d4aec1f5#
"... + {- HLINT ignore "Use zipFrom" -} ..."

Rebased. Dropped use of zipFrom and zipFromWith. Moved the zipFrom and zipFromWith ignores to .hlint.yaml. Dropped all changes to hls-tactics-plugin/**.

jneira commented:
Lets try merge this asap to avoid more conflicts

jhrcek commented:
FYI this commit seems to have broken the install script.
Doing `stack ./install.hs hls-8.8.4` works with previous commit, but not with this commit.

There's no clear error message:

```bash
$ stack ./install.hs hls-8.8.4
--  While building package hls-install-0.8.0.0 (scroll up to its section to see the error) using:
      /home/jhrcek/.stack/setup-exe-cache/x86_64-linux-tinfo6/Cabal-simple_mPHDZzAJ_2.4.0.1_ghc-8.6.5 --builddir=.stack-work/dist/x86_64-linux-tinfo6/Cabal-2.4.0.1 build lib:hls-install --ghc-options " -fdiagnostics-color=always"
    Process exited with code: ExitFailure 1
```

jneira commented on install/src/Env.hs:
```diff
@@ -54,7 +54,7 @@ findInstalledGhcs = do
     $ sortBy (comparing fst)
     -- nub by version. knownGhcs takes precedence.
-    $ nubBy ((==) `on` fst)
+    $ nubOrdBy (compare `on` fst)
```
Maybe this change has broken the install script??? It shouldn't but is the unique direct change in that component. 
@jhrcek please, could you try the install reverting this change manually?

jneira replied:
If the change broke it, hlint suggestion was very wrong :thinking:

jhrcek replied:
@jneira Just tried with latest master and I had to revert both this change (nubOrdBy back to nubBy) AND one more change done in `install/src/Print.hs` below. After reverting those two changes it started working again.

jhrcek replied:
I don't know if there's some kind of implicit import going on or something ,but those functions come from `extra` and I don't see them imported in either of the 2 modules.

jneira replied:
No implicit imports afaik, fortunately :slightly_smiling_face: 
The fix would be add those imports probably

jhrcek replied:
Sure, I'll do it later today.

peterwicksstringfield replied:
I guess this part of the code isn't tested by the CI then? Sorry about that.

@jneira thanks for the merge!
@jhrcek thanks for fixing it!

(I did not except this PR to be so fiddly and hard to merge.)

jneira replied:
yeah, the install script is not being tested, but it does not change frequently so 🤷
